apps/src/storage/dataBrowser/dataVisualizer/Snapshot.jsx

```diff
@@ -4,28 +4,76 @@ import PropTypes from 'prop-types';
 import {connect} from 'react-redux';
 import moment from 'moment';
 import msg from '@cdo/locale';
+import {svgToDataURI} from '@cdo/apps/imageUtils';
 import {html2canvas} from '@cdo/apps/util/htmlToCanvasWrapper';
 import BaseDialog from '@cdo/apps/templates/BaseDialog';
 import PendingButton from '@cdo/apps/templates/PendingButton';
+import {ChartType} from '../dataUtils';
 import * as dataStyles from '../dataStyles';
 
+const INITIAL_STATE = {
+  isSnapshotOpen: false,
+  isCopyPending: false,
+  isSavePending: false,
+  imageSrc: require('./placeholder.png')
+};
+
 class Snapshot extends React.Component {
   static propTypes = {
+    chartType: PropTypes.number.isRequired,
     chartTitle: PropTypes.string.isRequired,
     selectedOptions: PropTypes.string.isRequired,
     // Provided via Redux
     tableName: PropTypes.string.isRequired,
     projectName: PropTypes.string.isRequired
   };
 
-  state = {
-    isSnapshotOpen: false,
-    isCopyPending: false,
-    isSavePending: false
+  state = {...INITIAL_STATE};
+
+  componentDidMount() {
+    this.isMounted_ = true;
+  }
+
+  componentWillUnmount() {
+    this.isMounted_ = false;
+  }
+
+  handleOpen = () => {
+    this.setState({isSnapshotOpen: true});
+    this.getImageFromChart();
   };
 
-  handleOpen = () => this.setState({isSnapshotOpen: true});
-  handleClose = () => this.setState({isSnapshotOpen: false});
+  handleClose = () => this.setState(INITIAL_STATE);
+
+  getImageFromChart = () => {
+    switch (this.props.chartType) {
+      case ChartType.BAR_CHART:
+      case ChartType.HISTOGRAM:
+      case ChartType.SCATTER_PLOT:
+        this.getImageFromGoogleChart();
+        break;
+      case ChartType.CROSS_TAB:
+        this.getImageFromCrossTab();
+        break;
+      default:
+    }
+  };
+
+  getImageFromCrossTab = () => {};
+
+  getImageFromGoogleChart = () => {
+    const container = document.getElementById('googleChartContainer');
+    const svgList = container && container.querySelectorAll('svg');
+    const svg = svgList && svgList[0];
+    if (!svg) {
+      return;
+    }
+    svgToDataURI(svg, 'image/png', {renderer: 'native'}).then(imageURI => {
+      if (this.isMounted_) {
+        this.setState({imageSrc: imageURI});
+      }
```

Comment:
I've heard [isMounted is an antipattern](https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html) but have never been totally satisfied with the alternatives - cancellable promises are cool, but also have more boilerplate than I'd like.

No need to change anything, but I'd love to hear your thoughts on this.

Reply:
Oo thanks for pointing this out. I didn't know isMounted was an anti-pattern, I think I just modeled this off of similar structures elsewhere in our codebase. I will look into cancellable promises as a follow-up PR.